h/accounts/util.py

```diff
@@ -0,0 +1,74 @@
+"""
+Helpers for account forms
+"""
+
+import re
+
+from h._compat import urlparse
+
+def validate_url(url):
+    """
+    Validate an HTTP(S) URL as a link for a user's profile.
+
+    Helper for use with Colander that validates a URL provided by a user as a
+    link for their profile.
+
+    Returns the normalized URL if successfully parsed or raises a ValueError
+    otherwise.
+    """
+
+    # Minimal URL validation with urlparse. This is extremely lenient, we might
+    # want to use something like https://github.com/kvesteri/validators instead.
+    parsed_url = urlparse.urlparse(url)
+
+    if not parsed_url.scheme:
+        parsed_url = urlparse.urlparse('http://' + url)
+
+    if not re.match('https?', parsed_url.scheme):
+        raise ValueError('Links must have an "http" or "https" prefix')
+
+    if not parsed_url.netloc:
+        raise ValueError('Links must include a domain name')
+
+    return parsed_url.geturl()
+
+
+def orcid_checksum_digit(orcid):
+    """
+    Return the checksum digit for an ORCID identifier.
+
+    Translated from the example ISO 7064 checksum implementation at
+    http://support.orcid.org/knowledgebase/articles/116780-structure-of-the-orcid-identifier
+    """
+    total = 0
+    digits = [int(ch) for ch in orcid if ch.isdigit()]
```

Comment:
I'm not sure `isdigit()` is quite what you're looking for here. In Python 3:

```
>>> '²'.isdigit()
True
```

(That's a Unicode superscript character.)

Given the hard restrictions on valid inputs to this function, I'd be inclined to do something like this in `validate_orcid`:

```python
orcid_normalised = orcid.replace('-', '')

if not _validate_iso7064_mod_11_2(orcid_normalised):
    raise ValueError('{} is not a valid ORCID'.format(orcid))
```

and then define

```python
def _validate_iso7064_mod_11_2(value):
    check = 0
    for n in value:
        check = (2 * check + int(10 if n == 'X' else n)) % 11
    return check == 1
```